[Mark Biju George] iP

markbiju opened the pull request:
# DukePro

> "The best time to plant a tree was 20 years ago. The second best time is now.” - Chinese Proverb ([source](https://goleansixsigma.com/the-best-time-to-plant-a-tree-was-20-years-ago-the-second-best-time-is-now-chinese-proverb/#:~:text=The%20second%20best%20time%20is%20now.%E2%80%9D%20%E2%80%93%20Chinese%20Proverb))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1.  download it from [here](https://github.com/markbiju/ip/releases/download/A-Jar/main.jar)
2. open a command line prompt
3. cd to where the downloaded file is(e.g cd \User\John\Desktop)
4. type this command: java -jar main.jar
5. add your tasks.
6. let it manage your tasks for you 😏

And it is **FREE!**

Features

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Here's a sneak peek into what's going on behind the scenes😏
```
public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

markbiju16 committed:
Update Duke.java

markbiju16 committed:
Update Duke.java

markbiju16 committed:
Update Duke.java

markbiju16 committed:
Level-4

markbiju16 committed:
Level-4 upd

markbiju16 committed:
A-TextUiTesting

markbiju16 committed:
Level-5

markbiju16 committed:
Update Duke.java

LowJiaHao99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,124 @@
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+    public static void main(String[] args) throws DukeException {
+        Scanner myObj = new Scanner(System.in);
```
Would "scanner" be a better variable name here?

LowJiaHao99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,124 @@
+        int counter = 0;
+        ArrayList<Task> tasks = new ArrayList<Task>();
+
+        System.out.println("Hello! I'm Duke");
```
Would you consider creating a constant final String for these messages to be printed?

LowJiaHao99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,124 @@
+            if (input.equals("bye")) {
+                System.out.println("Bye. Hope to see you again soon!");
+                break;
+            } else if (input.equals("list")){
```
As per java Coding Standards

>Java reserved words should be followed by a white space.

Perhaps you can add a white space between the condition and the open parenthesis to be consistent with the rest of the code.

I notice many similar issues throughout the rest of the code.

LowJiaHao99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,124 @@
+                    System.out.println(currentTask);
+                } catch (StringIndexOutOfBoundsException e) {
+                    throw new DukeException("☹ OOPS!!! Please tell me which task you want me to mark as not done!");
+
```
Perhaps this might be a mistake, do remember to leave out empty lines within a logic block.

I see such mistake appearing occasionally in the rest of the code.

LowJiaHao99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,124 @@
-        System.out.println("Hello from\n" + logo);
+    public static void main(String[] args) throws DukeException {
+        Scanner myObj = new Scanner(System.in);
+        int counter = 0;
```
How about a more intuitive name here? Suggestion: "numOfTasksInList".

LowJiaHao99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,124 @@
+                } catch (StringIndexOutOfBoundsException e) {
+                    throw new DukeException("☹ OOPS!!! Please tell me which task to mark as done!");
+                }
+
```
Are the empty lines used to differentiate between the next condition in the user input processing? This is a pretty good style as it notifies readers/reviewers of the change in logic.

LowJiaHao99 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+        this.isDone = false;
+    }
+
+    public String getStatusIcon() {
```
I like the way you name this method. It is short, direct and natural. 

Much better than getIsDone() which is what I named mine. 😅

LowJiaHao99 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+    }
+
+    @Override
+    public String toString() {
```
I like the use of toString() to print details of the task. The method name itself is already specific to its action and it is a built-in method in all java objects.

LowJiaHao99 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
```
Perhaps this variable name can be more specific? One might mistake 'by' as 'added by' instead of 'complete by'.

LowJiaHao99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,124 @@
+                    }
+                } catch(ArrayIndexOutOfBoundsException e) {
+                    throw new DukeException("☹ OOPS!!! Please provide a timing for your event.");
+                }
```
Perhaps this might have been missed during checks, did you made a mistake and left out a empty white line to separate the if-else logic blocks?

LowJiaHao99 reviewed (commented):
Overall, this is a pretty good effort in following Java Coding Standards closely. The code is pretty clean and simple and it helps to improve readability. Some names of the variable can be improved to further enhance it. There are also some inconsistency, that needs addressing, in regards to the way you organize your code.

Other than that, good job!! 👍

markbiju16 committed:
Level-7

markbiju16 committed:
Level-8

markbiju16 committed:
Merge branch 'branch-Level-8'

paing-oo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,182 @@
+            f.createNewFile();
+        }
+
+        while (true) {
```
Perhaps, parsing user inputs could be done in another class called Parser for more abstraction and OOP.

paing-oo replied:
Perhaps, this code block is too long, could work towards better abstraction overall.

paing-oo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,182 @@
+                    int toBeMarked = Integer.parseInt(input.substring(5));
+                    Task currentTask = tasks.get(toBeMarked - 1);
+                    currentTask.markAsDone();
+                    System.out.println("Nice! I've marked this task as done:");
```
Perhaps, printing of messages could be done in another Ui class for better abstraction and OOP.

markbiju16 committed:
A-MoreOOP

markbiju16 committed:
update A-MoreOOP

markbiju16 committed:
Delete AddCommand.java

markbiju16 committed:
A-Packages

markbiju16 committed:
A-JUnit

markbiju16 committed:
Delete file.txt

markbiju16 committed:
A-Jar

markbiju16 committed:
branch-A-JavaDoc

markbiju16 committed:
branch-A-CodingStandard

markbiju16 committed:
Level-9

markbiju16 committed:
Merge branch 'branch-A-CodingStandard'

markbiju16 committed:
Merge branch 'branch-Level-9'

markbiju16 committed:
Merge branch 'add-gradle-support'

markbiju16 committed:
Level-10

markbiju16 committed:
Update Storage.java

markbiju16 committed:
update Ui

markbiju16 committed:
Update tasks.txt

markbiju16 committed:
Update build.gradle

markbiju16 committed:
update A-Jar

markbiju16 committed:
update Unmark

markbiju16 committed:
Update find

markbiju16 committed:
Update delete

markbiju16 committed:
Add assertions

Parsing and storage may occur differently if there are more spaces/unexpected input

Assertions are needed to test whether parsing and storage is working properly

Add assertions in parts of storage and parser code to enforce testing.

Let's observe the results of these assertions using different inputs

markbiju16 committed:
Update Ui.java

markbiju16 committed:
Merge branch 'branch-Level-10'

markbiju16 committed:
Improve code quality

Some lines are too long.

They need to be shortened to improve readability.

Line breaks are placed such that longer lines of code are broken down into shorter segments

markbiju16 committed:
Update Deadline.java

markbiju16 committed:
Update Deadline.java

markbiju committed:
Merge branch 'master' into branch-A-Assertions

markbiju committed:
Merge pull request #2 from markbiju/branch-A-Assertions

Add assertions

markbiju committed:
Merge branch 'master' into branch-A-CodeQuality

markbiju committed:
Merge pull request #3 from markbiju/branch-A-CodeQuality

Branch A-CodeQuality

markbiju16 committed:
Update deadline saving

markbiju16 committed:
Merge branch 'master' of https://github.com/markbiju/ip

markbiju16 committed:
Update deadline saving

markbiju16 committed:
BCD-Extension

markbiju16 committed:
Update README.md

markbiju16 committed:
Ui.png

markbiju16 committed:
Update README.md

markbiju16 committed:
Update README.md

markbiju16 committed:
Update README.md

markbiju16 committed:
Update SnoozeCommand.java

markbiju committed:
Update README.md

markbiju16 committed:
Update Parser.java

markbiju16 committed:
Update Parser.java
